### PR TITLE
My cart Page

### PR DESCRIPTION
**Pull request by Chandrika-Rani-Infrrd**
added:routing to product-list to navigate to my cart
updated :product list spec file with few test case
updated:mycart file to get and display selected products
added:quantity increment and decrement,delete,total amount to my cart page
updated :mycart spec file with few test cases
updated:product json with id and quantity property

### CODE REVIEW

File: src/app/mycart/my-cart.component.html [hunk 1]
```diff
@@ -1 +1,24 @@
...
+            <button (click)="increment(item)" >+</button>
+        </div>
+        <div><b>Price : </b>{{item.price}}</div>
+        <div><b>Total Price : </b>{{total(item.quantity,item.price)}}</div>
+    </div>  
+    <div class="action">
+        <button class="delete" (click)="productToDelete(item.id)">Delete</button>
+    </div>
+</div> 
+<div class="total">Total Amount : {{sum()}}</div>
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.html]: You shouldn't call functions from the template unless they are based on an event, like click, scroll etc. Template functions will get triggered on every change detection cycle. Use a class property instead.

File: src/app/mycart/my-cart.component.html [hunk 2]
```diff
@@ -1 +1,24 @@
...
+<div class="table-body" *ngFor="let item of cartValues">
+    <div class="table-details">
+        <div><b>Product Name : </b>{{item.name}}</div>
+        <div class="quantity"><b>Quantity : </b>
+            <button (click)="decrement(item)">-</button>
+                {{item.quantity}}
+            <button (click)="increment(item)" >+</button>
+        </div>
+        <div><b>Price : </b>{{item.price}}</div>
+        <div><b>Total Price : </b>{{total(item.quantity,item.price)}}</div>
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.html]: Don't use a function, use a property or a pipe instead

Comment (lloydaf) [reply src/app/mycart/my-cart.component.html]: Reason explained below 👇

File: src/app/mycart/my-cart.component.spec.ts [hunk 3]
```diff
@@ -22,4 +31,57 @@ describe('MycartComponent', () => {
...
     expect(component).toBeTruthy();
   });
+
+  it("does more shopping btn navigates to product list",
+  async(inject([Router, Location], (router: Router, location: Location) => {
+  let fixture = TestBed.createComponent(MyCartComponent);
+  fixture.detectChanges(); 
+
+  router.navigate(['product-list']).then(() => {
+    expect(location.path()).toBe('/product-list');
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.spec.ts]: 💯

File: src/app/mycart/my-cart.component.spec.ts [hunk 4]
```diff
@@ -22,4 +31,57 @@ describe('MycartComponent', () => {
...
+  })  
+
+  it("service should be created", () => {
+    const service:ProductsService=TestBed.get(ProductsService)
+    expect(service).toBeTruthy();
+  });
+  
+  it("should have getProducts method",()=>{
+    const service:ProductsService=TestBed.get(ProductsService)
+    expect(service.getProducts).toBeTruthy();
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.spec.ts]: 💯

File: src/app/mycart/my-cart.component.ts [hunk 5]
```diff
@@ -1,15 +1,82 @@
...
   selector: 'app-my-cart',
   templateUrl: './my-cart.component.html',
   styleUrls: ['./my-cart.component.css']
 })
 export class MyCartComponent implements OnInit {
+  name:string;
+  productsList:any=[];
+  myCart:any=[];
+  cartValues=[];
+  totalAmount:number=0;
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you needn't really specify the type here. Javascript can imply the type from the value

File: src/app/mycart/my-cart.component.ts [hunk 6]
```diff
@@ -1,15 +1,82 @@
...
+    this.name=this.route.snapshot.paramMap.get('name');
+    this.productsService.getProducts().subscribe(item=>{
+      this.productsList=item;
+
+      this.cartValues=JSON.parse(localStorage.getItem('cartValue')) || [];
+      console.log("first",this.cartValues); 
+
+      for(let i=0;i<this.productsList.length;i++){
+        if(this.productsList[i].name===this.name && 
+          !this.cartValues.map(item=>item.name).includes(this.name)){
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: You shouldn't use a loop within a loop. `O(n^2)` time complexity.

Comment (lloydaf) [reply src/app/mycart/my-cart.component.ts]: To resolve this issue, make ProductsList a set or a map. You don't have to do it that way, just a suggestion, upto you

File: src/app/mycart/my-cart.component.ts [hunk 7]
```diff
@@ -1,15 +1,82 @@
...
+            this.cartValues.push(this.productsList[i]);
+            console.log("after push",this.productsList[i]);
+            localStorage.setItem('cartValue',JSON.stringify(this.cartValues));
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you shouldn't really do this within every iteration. Expensive operation right?

File: src/app/mycart/my-cart.component.ts [hunk 8]
```diff
@@ -1,15 +1,82 @@
...
+            let cart=JSON.parse(localStorage.getItem('cartValue'));
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: You shouldn't probably do this either

File: src/app/mycart/my-cart.component.ts (hunk 7 above)
Comment (Chandrika-Rani-Infrrd) [reply src/app/mycart/my-cart.component.ts]: can i create a service and then use only cartvalues in mycart file?

File: src/app/mycart/my-cart.component.ts [hunk 9]
```diff
@@ -1,15 +1,82 @@
...
+            console.log("second",cart);         
+        }
+      }           
+    })
+  }
+
+  productToDelete(value){
+    console.log("item"+value);
+    this.cartValues.splice(this.cartValues.findIndex(item=>item.id==value),1);
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you can just use a filter instead, than using findIndex within splice. Try to keep your code as functional as possible, with minimum mutations. you can read about it here https://medium.com/@cristiansalcescu/learn-immutability-with-javascript-6a67e4a48d7f

File: src/app/mycart/my-cart.component.ts [hunk 10]
```diff
@@ -1,15 +1,82 @@
...
+    localStorage.setItem('cartValue',JSON.stringify(this.cartValues));
+  }
+
+  total(quantity,price):number{
+    let total=(quantity*price);
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you don't need a function for this

File: src/app/mycart/my-cart.component.ts [hunk 11]
```diff
@@ -1,15 +1,82 @@
...
+      item.quantity=1;
+    else
+       item.quantity-=1; 
+    localStorage.setItem('cartValue',JSON.stringify(this.cartValues));
+    return item.quantity;
+  }
+
+  sum():number{
+    let sum=0;
+    for(let i in this.cartValues){
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you can use `.reduce` instead of a for loop here

File: src/app/mycart/my-cart.component.ts [hunk 12]
```diff
@@ -1,15 +1,80 @@
...
+  checkDeletedItem: any;
+  deletedItemPresent: boolean;
 
-  constructor() { }
+  constructor(private router:Router,
+              private route:ActivatedRoute,
+              private productService:ProductsService){}
 
   ngOnInit(): void {
+    if(localStorage.getItem('deleteditem').length>0){
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: use a variable instead of a hard coded string value

File: src/app/mycart/over-all-total.pipe.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+  pure:false
+})
+export class OverAllTotalPipe implements PipeTransform {
+
+  transform(value:any,products:any){
+    return this.overAllTotal(value,products);
+  }
+
+  overAllTotal(sum,products){
+    sum=products.reduce((acc,item)=>acc+(item.quantity*item.price),0);
```
Comment (lloydaf) [inline src/app/mycart/over-all-total.pipe.ts]: what is the purpose of passing sum as an argument if you are going to overwrite the value?

File: src/app/app-routing.module.ts [hunk 14]
```diff
@@ -4,16 +4,15 @@ import { Routes, RouterModule } from '@angular/router';
...
-
 export const routes: Routes = [
-  {path:'',redirectTo:'product-list',pathMatch:'full'},
-  {path:'product-list',component:ProductListComponent},
-  {path:'mycart',component:MyCartComponent},
-  {path:'products/:id',component:ProductListComponent},
+  { path: '', redirectTo: 'product-list', pathMatch: 'full' },
+  { path: 'product-list', component: ProductListComponent },
+  { path: 'mycart', component: MyCartComponent },
+  { path: 'mycart/:name', component: MyCartComponent },
```
Comment (lloydaf) [inline src/app/app-routing.module.ts]: remove this `:name` parameter

File: src/app/mycart/my-cart.component.spec.ts [hunk 15]
```diff
@@ -22,4 +40,60 @@ describe('MycartComponent', () => {
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('create an instance for pipe', () => {
+    const pipe = new OverAllTotalPipe();
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.spec.ts:44]: 💯

File: src/app/mycart/my-cart.component.ts [hunk 16]
```diff
@@ -1,15 +1,60 @@
...
 @Component({
   selector: 'app-my-cart',
   templateUrl: './my-cart.component.html',
-  styleUrls: ['./my-cart.component.css']
+  styleUrls: ['./my-cart.component.css'],
+  providers: [DeleteComponent],
 })
 export class MyCartComponent implements OnInit {
+  myCartValues = [];
+  name: string;
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: you don't need this

File: src/app/mycart/my-cart.component.ts [hunk 17]
```diff
@@ -1,15 +1,60 @@
...
-  constructor() { }
+  constructor(
+    private router: Router,
+    private mycartService: MyCartService,
+    private route: ActivatedRoute,
+    private deleteItem: DeleteComponent
+  ) {}
 
   ngOnInit(): void {
+    this.name = this.route.snapshot.paramMap.get('name');
```
Comment (lloydaf) [inline src/app/mycart/my-cart.component.ts]: don't need this

File: src/app/mycart/my-cart.service.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class MyCartService {
+
+  private selectedItem:any;
+
```
Comment (lloydaf) [inline src/app/mycart/my-cart.service.ts]: addItemToCart(),
removeItemFromCart(),
increment, decrement

**Pull request merged**